### PR TITLE
Add causal survival forest

### PR DESCRIPTION
**Pull request by erikcs**
#### Overview

* Add `causal_survival_forest(X, Y, W, D)` consisting of
   - `CausalSurvivalRelabelingStrategy` 
  - `CausalSurvivalSplittingRule`
    This is exactly the same as `InstrumentalPredictionRule` but with the additional requirement that there are at least `min_child_size` number of failures in each child node. Details: we use `InstrumentalPredictionStrategy`'s "`W`-constraint" on `W - W.hat`, with the minimum child criterion being `alpha` * "effective size parent". The added `D`-constraint is that there are at least `alpha` * "number of samples in parent" number of failures in each child. (a future simplification is pending #642)
   - `CausalSurvivalPredictionStrategy`

#### Details

- More computations in R wrapper code than earlier forests - more work for potential Python wrappers. 

Since this feature will initially be experimental and subject to change, some details can be changed later, such as:
- All notation consistent with paper or rest of package? (for example, use a different word than `tau` for maximum follow-up time?)
- The way optional nuisance parameters are passed may be a bit finicky: we require `S.hat` and `C.hat` to both correspond to the same grid `U`, since they enter computations where both are indexed by the same time value, but `S0.hat` and `S1.hat` can be fitted on an arbitrary grid since that issue does not arise with these objects. In the future we may just fit everything on the same grid `U` for simplicity.

#### Follow up PR:
- Integration with summary functions
- Best linear projection
- Tuning? (Prediction: `debiased.error` and `excess.error`)
- Update causal_benchmark
- Add paper code to `experiments/`

### CODE REVIEW

File: r-package/grf/tests/testthat/data/test_causal_survival_forest_1.csv [hunk 1]
```diff
@@ -0,0 +1,500 @@
+0.444564019146147
```
Comment (jtibshirani) [inline r-package/grf/tests/testthat/data/test_causal_survival_forest_1.csv:1]: It would be good to make a note somewhere about how this data was generated. And as a small comment, the file names could be more descriptive file names instead of using 1, 12, ...

File: r-package/grf/tests/testthat/test_causal_survival_forest.R [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+  csf.mia <- causal_survival_forest(X.mia, data$Y, data$W, data$D, num.trees = 500)
+
+  mse.oob.diff <- mean((predict(csf)$predictions - predict(csf.mia)$predictions)^2)
+  mse.diff <- mean((predict(csf, data$X)$predictions - predict(csf.mia, X.mia)$predictions)^2)
+
+  expect_equal(mse.oob.diff, 0, tol = 0.001)
+  expect_equal(mse.diff, 0, tol = 0.001)
+})
+
+# This characterization test locks in behavior for the default causal survival forest.
```
Comment (jtibshirani) [inline r-package/grf/tests/testthat/test_causal_survival_forest.R:149]: Good idea to add this!

File: core/test/forest/CausalSurvivalForestTest.cpp [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+#include "forest/ForestPredictors.h"
+#include "forest/ForestTrainer.h"
+#include "forest/ForestTrainers.h"
+#include "utilities/ForestTestUtilities.h"
+
+#include "catch.hpp"
+
+using namespace grf;
+
+TEST_CASE("causal survival forests give positive variance estimates", "[causal survival]") {
```
Comment (jtibshirani) [inline core/test/forest/CausalSurvivalForestTest.cpp:29]: When implementing survival forests, we added `SurvivalSplittingRuleTest` and `SurvivalPredictionStrategyTest`. How feasible would it be to add this sort of unit test for causal survival forests?

File: core/src/splitting/CausalSurvivalSplittingRule.cpp [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+}
+
+bool CausalSurvivalSplittingRule::find_best_split(const Data& data,
+                                                  size_t node,
+                                                  const std::vector<size_t>& possible_split_vars,
+                                                  const std::vector<double>& responses_by_sample,
+                                                  const std::vector<std::vector<size_t>>& samples,
+                                                  std::vector<size_t>& split_vars,
+                                                  std::vector<double>& split_values,
+                                                  std::vector<bool>& send_missing_left) {
```
Comment (jtibshirani) [inline core/src/splitting/CausalSurvivalSplittingRule.cpp:72]: The other splitting rules have tests in `SplitRuleInvarianceTest` for missing values -- we could add a case for this new rule.

File: r-package/grf/R/causal_survival_forest.R [hunk 5]
```diff
@@ -0,0 +1,758 @@
...
+#' T(1) and T(0) are potental outcomes corresponding to the two possible
+#' treatment states.
+#'
+#' @param X The covariates.
+#' @param Y The event time (may be negative).
+#' @param W The treatment assignment (must be a binary vector with no NAs).
+#' @param D The event type (0: censoring, 1: failure).
+#' @param W.hat Estimates of the treatment propensities E[W | Xi]. If W.hat = NULL,
+#'              these are estimated using a separate regression forest. Default is NULL.
+#' @param S1.hat Estimates of the conditional survival function S(t, x, 1) = P[Y > t | X = x, W = 1].
```
Comment (jtibshirani) [inline r-package/grf/R/causal_survival_forest.R:16]: Would we plausibly expect users to pass in values for `S1.hat`, `S0.hat`, `C.hat`, `lambda.C.hat`? If not, we could avoid making them configurable to reduce the API surface.

File: core/src/splitting/CausalSurvivalSplittingRule.cpp (hunk 4 above)
Comment (erikcs) [reply core/src/splitting/CausalSurvivalSplittingRule.cpp:72]: Yes, I've added one now!

**Comment by erikcs** [comment]
> Could you expand on what you mean here?

Yes, it is only that `S1.hat` and `S0.hat` are lists (first entry is the survival matrix and second the corresponding grid), while the rest are just the survival matrices (but constrained to correspond to the full `Y`-grid, which `S1.hat` and `S0.hat` do not necessarily need)

File: r-package/grf/R/causal_survival_forest.R (hunk 5 above)
Comment (erikcs) [reply r-package/grf/R/causal_survival_forest.R:16]: It is for consistency with other forests that has nuisance parameters, like `Y.hat` and `W.hat` the user may want to fine tune (change min.node.size etc), or for example, a user can threshold low inverse probability weighting of censoring weights by doing
```
sf.censor = survival_forest(X, Y, 1 - D, prediction.type = "Nelson-Aalen")
C.hat = predict(sf.censor, failure.times = sort(unique(Y)))$predictions
C.hat[C.hat < 0.01] = 0.01
cs.forest = causal_survival_forest(X, Y, W, D,
                                   C.hat = C.hat)
```

and for ` lambda.C.hat`, a user may fine-tune this by calculating derivates from a spline fit instead of the default forward difference.

File: core/test/forest/CausalSurvivalForestTest.cpp (hunk 3 above)
Comment (erikcs) [reply core/test/forest/CausalSurvivalForestTest.cpp:29]: I think it may be trickier with direct tests here because `SurvivalSplitting` had ground truth to compare with (log-rank test statistic for splitting, and Kaplan-Meier estimate for prediction) and this is practically the same as Instrumental/Causal forest where there are indirect test that behavior is reasonable

File: r-package/grf/tests/testthat/data/README.md [hunk 6]
```diff
@@ -0,0 +1,36 @@
+# Datasets
+
+The datasets is this directory were generated through the following R scripts:
```
Comment (jtibshirani) [inline r-package/grf/tests/testthat/data/README.md:3]: I'm sorry about my incorrect suggestion earlier -- I read too fast and didn't see that all of these CSV files are simply used in characterization tests. This should have been obvious since they all end in `_predictions`!

We could omit the README and go with what you had before then, since it nicely matches the C++ characterization tests.

File: core/test/forest/CausalSurvivalForestTest.cpp (hunk 3 above)
Comment (jtibshirani) [reply core/test/forest/CausalSurvivalForestTest.cpp:29]: That makes sense. I'm okay with the level of test coverage, especially if we have someone test it manually and 'lock in' the behavior with characterization tests. But just to brainstorm, if you're looking for more testing ideas: we could perhaps unit test some of the helper functions like `compute_eta` and the nuisance component calculations?

File: r-package/grf/R/causal_survival_forest.R (hunk 5 above)
Comment (jtibshirani) [reply r-package/grf/R/causal_survival_forest.R:16]: Okay, this makes sense. I also caught up with @swager and he confirms that it's helpful to be able to pass these in (even just for performance reasons -- you can calculate nuisance parameters once, then try several different forests).

File: core/test/forest/CausalSurvivalForestTest.cpp (hunk 3 above)
Comment (erikcs) [reply core/test/forest/CausalSurvivalForestTest.cpp:29]: Oki. Computing eta is practically just a bunch of tedious algebra: Screen Shot 2020-08-19 at 22 49 58

it involved taking care of a handful of edge cases which is locked in with the current characterization tests. I thought about tests for it, but did not come up with much, like plugging in oracle components for a simple DGP and seeing that the output is reasonable - it is already implicitly verified in the final simulation results in the paper.

File: core/src/prediction/CausalSurvivalPredictionStrategy.cpp [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+
+size_t CausalSurvivalPredictionStrategy::prediction_length() const {
+    return 1;
+}
+
+std::vector<double> CausalSurvivalPredictionStrategy::predict(const std::vector<double>& average) const {
+  return { average.at(NUMERATOR) / average.at(DENOMINATOR) };
+}
+
+std::vector<double> CausalSurvivalPredictionStrategy::compute_variance(
```
[the remark this answers is not in the document]
Comment (erikcs) [reply core/src/prediction/CausalSurvivalPredictionStrategy.cpp:40]: I am sorry, did you mean 0cf5083 which I corrected now, or something else? I followed the indent from [InstrumentalPredictionStrategy](https://github.com/erikcs/grf/blob/add-causal-sf/core/src/prediction/InstrumentalPredictionStrategy.cpp#L48) where the function parameters where indented 4 spaces?

File: r-package/grf/tests/testthat/data/README.md (hunk 6 above)
Comment (erikcs) [reply r-package/grf/tests/testthat/data/README.md]: Oki, I've deleted this file and added the details to the R test file instead

File: core/src/prediction/CausalSurvivalPredictionStrategy.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (erikcs) [reply core/src/prediction/CausalSurvivalPredictionStrategy.cpp:40]: Sorry about that, it should be fixed in 469805e

File: core/test/forest/CausalSurvivalForestTest.cpp (hunk 3 above)
Comment (erikcs) [reply core/test/forest/CausalSurvivalForestTest.cpp:29]: Also: @yifan-cui and I have done sanity checks by verifying reasonable output with DGPs over the course of this PR.

@jmzeng has also kindly beta tested the package, and as we are marking this first release "Experimental" we can incorporate improvements and suggestions as more people try it out.

Comment (jtibshirani) [reply core/test/forest/CausalSurvivalForestTest.cpp:29]: 👍  it sounds like we really have confidence in correctness.

**Comment by erikcs** [comment]
Would anyone object to amend the `causal_survival_forest` optional parameter API:

Instead of

```
S0.hat : a list of ...
S1.hat:  a list of ...
```

Simplify it to:

```
E0.hat: An n-vector of estimates of E[Y | X, W = 1], the expected survival time conditioning on being treated. If NULL (default), then estimated with an S learner using a survival forest. (...)
E1.hat: " "
```

**Comment by yifan-cui** [comment]
@erikcs Looks good to me.
